### DISCUSSION TITLE
WindTerm 2.4.0 Prerelease 1

### DISCUSSION THREAD

**Discussion by kingToolbox** [kingToolbox/WindTerm#606] [General]
# Note:

**WindTerm 2.4.0 changes the upgrade method to simplify subsequent upgrades or cloud synchronization. Please read the `Upgrade` to learn more.**

# Usage:
### Windows

|   Version | Architecture |  Operating Systems |
| --------- | -------------- | --------------------- |
| WindTerm_2.4.0_Prerelease_1_Windows_Portable_x86_64 | x86_64 | Windows 7, Windows 8, Windows 10, Windows 11 |

#### Versions before Version 2.4.0

1. Start `WindTerm` and select the `profiles directory` and **quit**.
2. Replace the new `.wind/profiles` folder with the old `profiles` folder.
3. Restart `WindTerm` and all the sessions and configuration will be restored.

#### Versions after Version 2.4.0

1. Start `WindTerm` and select the `profiles directory` containing the old `.wind` folder.
2. It is recommended to choose the `Home Directory` to facilitate subsequent upgrades.

If the `.wind` folder needs to be moved to another system:
1. Quit `WindTerm`.
4. Replace the new `.wind` folder with the old `.wind` folder.
5. Restart `WindTerm` and all the sessions and configuration will be restored.

The **.wind** folder is cross-platform. 

# Known Issues:
- [ZMODEM] **Only support file names contains ASCII characters in Windows, otherwise the file name of the transfered file will be garbled after transmission, but it does not affect the correctness of the file content.** Linux and MacOS version does not have this problem.

# Changes in Version 2.4.0 Prerelease 1 (2022-3-20):

# Translation:
- [FRENCH] Update the language file, contributed by **EvoWebFrance@github** and **LuxNegra@github**. #499 

# Features:
- [APP] **Supports specifying the path of the `.wind` profiles folder to facilitate subsequent upgrades and cloud synchronization.** #161 #411 #438 #529 #537 #541 #601 
- [GUI] Add `Anything Palette` and `Tunnels` buttons to the menu bar.
- [TAB] **Supports using `Alt+1` to `Alt+5` to select the first to fifth tab, and `Alt+0` to select the last tab.** #82 #86 #420 
- [LOG] The log file name and content templates support including `%z - Time Zone`.
- [SSH] **Supports independent tunnels opening, stopping and management.** #189 #194 #373 #444 
- [SSH] When the listening port of a `tunnel` is set to 0, `dynamic port allocation` is supported.
- [SESSION] **Supports setting the background color of the tab of a session.**
- [SESSION] **Supports using the shortcut key `Ctrl+D` to close the unconnected or disconnected session.** #397 #526 
- [SESSION] **Supports new proxy type `ProxyCommand`.** #557 
- [SESSION] Supports importing "ProxyTelnetCommand" field from PuTTY configuration.
- [SESSION] **Supports clearing the screen locally.** #335 #447 #530 #553 #571 
- [SCHEME] **Added `dumb` system to support highlighting, folding, outlining and `clearing the screen locally` of dumb devices.**
- [SCHEME] Supports highlighting `IPv6` and `MAC`.

# Improvements:
- [GUI] Simplify the menu items in the menu bar to provide more screen space.
- [GUI] The pop-up widget of the `More` button in the toolbar is modified to a menu.
- [TAB] Even when only one tab is open, the activation mark of the tab is displayed.
- [TAB] Improve the appearance of a tab by increasing the transparency of the background color and reducing its drawing area.
- [TRANSLATION] Automatically handle `mnemonic characters` in strings to be translated.

# Bugs:
- [GUI] When using the shortcut key `Ctrl+Shift+W` to close a tab, the confirm dialog will not pop up.
- [GUI] The text in the input box is sometimes displayed in gray as placeholder text.
- [GUI] After modifying the name or icon of a session, reconnecting the session did not use the new name and icon correctly.
- [GUI] When the bottom dock is expanded, the tooltip of the `Expand` button should be `Shrink`, not `Expand`.
- [TAB] The background color of the close button in the tab of the `Shell Pane` is not transparent.
- [LOG] **Blank lines in the log are incorrectly logged as `%v`.** #556
- [SESSION] When modifying `DataType` of a `Tcp` Session to `text` and saving, the value is incorrectly saved as `binary`.
- [SESSION] **When the session is disconnected, some cached text may not be correctly output to the screen.**
- [SESSION] **When the session is logged out, no `Remote channel is closed` is reported.**

<hr /><em>This discussion was created from the release <a href='https://github.com/kingToolbox/WindTerm/releases/tag/2.4-prerelease'>WindTerm 2.4.0 Prerelease 1</a>.</em>

**Comment by kingToolbox** [comment]
This is the first prerelease of WindTerm 2.4.0 version. Thanks to everyone who cares and helps, and a special thanks to [EvoWebFrance](https://github.com/EvoWebFrance) and [LuxNegra](https://github.com/LuxNegra) for the great French translations. :heart: 

Because I have some personal matters to take care of recently, the progress of development has been delayed, especially many `issues` have not been processed in time. I am really sorry for this. I will deal with `issues` as soon as possible and continue to release new prerelease versions. Please stay tuned, thank you.

**Reply by eugenov** [reply]
Great release! It looks like next major version update!

**Comment by SadarSSI** [comment]
Great ! the proxycommand option is promising! ;-)) there are still a few things to fix but a basic connection works [without evaluation of windows variables nor verbose output which causes problems]

The tunnel option does not work well
- the local tunnel stops after a few seconds / minutes
- the dynamic tunnel does not work (netstat -an see the open port, but impossible to connect to it)

Could you consider providing a field to put custom ssh values like -D xxx, -NL, etc?

**Comment by hisune** [comment]
Why does window switching only support `Alt+1` to `Alt+5` ? It would be great if it could support  `Alt+1` to `Alt+9`, I usually need to open a lot of windows.

**Comment by kingToolbox** [comment]
If you need, please open the file `WindTerm_installed_path/global/wind.keymaps` and add:
``` json
  { "keys": "<Alt+6>",
    "script": "() => { window.activeKit.activateView(6) }"
  },
  { "keys": "<Alt+7>",
    "script": "() => { window.activeKit.activateView(7) }"
  },
  { "keys": "<Alt+8>",
    "script": "() => { window.activeKit.activateView(8) }"
  },
  { "keys": "<Alt+9>",
    "script": "() => { window.activeKit.activateView(9) }"
  },
```
I currently have these keys reserved for future use in switching `kits`, `WindTerm` is actually only a part of the entire application, in addition to [WindEdit](https://github.com/kingToolbox/Windedit/) and so on, You can download [WindEdit 1.0.0 Prerelease 1](https://github.com/kingToolbox/WindEdit/releases/tag/1.0-prerelease) to have a look.
``` json
  { "keys": "<Alt+6>",
    "action": "Window.ActivateEditor" // Activate the editor kit - WindEdit
  },
  { "keys": "<Alt+7>",
    "action": "Window.ActivateTerminal" // Activate the terminal kit - WindTerm
  },
  { "keys": "<Alt+8>",
    "action": "Window.ActivateFiler"
  },
  { "keys": "<Alt+9>",
    "action": "Window.ActivateBrowser"
  }
```

**Comment by mizzuri** [comment]
Regarding the profile directory for WindTerm 2.4.0, it seems that WindTerm does not ask for the profile location again after the profile directory is deleted.

For example:

- Launch WindTerm for the first time, and it will ask where to save the profile directory.
- Quit WindTerm and delete the .wind directory.
- Relaunch WindTerm for the second time, but it will NOT ask where to save the profile directory this time. Instead, it will remember the previous location and create the .wind directory automatically. (e.g. At ~/.config/.wind)

Is there a file or a registry entry to edit to change the profile location?
Thank you.

**Reply by kingToolbox** [reply]
On startup, `WindTerm` will read the path of `profiles` from the `WindTerm_installed_path/profiles.config` file. And the path of `profiles` will only be asked again if the `profiles.config` file does not exist or the path of `profiles` does not exist. So you can reconfigure the path by deleting the `profiles.config` file.

In the upcoming `WindTerm_2.4.0_Prerelease_2` version, you can easily reconfigure the path by opening the `Choose Profiles Directory` dialog again via the menu item `Menubar - Menu - Preferences - Profiles Directory`.

BTW:
So far, `WindTerm` is still a portable software, so it will not write the registry.

**Comment by eugenov** [comment]
Hello!
How can I use Free type mode in alternative buffer without Alt modificator? It seems like I directly move cursor with mouse only in command line. In vim I can do it only with Alt pressed. Is it ok or I missed smth?
Thank you!
UPD: term set to xterm-256color

**Reply by kingToolbox** [reply]
Oh, that is how it was designed.

Although it is entirely possible to move the cursor directly with the mouse in `vi`. But once you switch to a program such as `less /etc/services`, it will cause many errors. Because for each click, `WindTerm` will calculate the control sequences to move the cursor from the current position to the specified position and send it to the server. And programs like `less` obviously don't handle these conrtol sequences correctly. Since there is currently no way for `WindTerm` to detect what the program running in the server is, the `Alt` modifier can only be added.

There is no such problem on the main screen, because `WindTerm` can recognize the start and end positions of the prompt line, and when the click position is not in it, the click will be ignored. Besides, even if the control sequence is sent by mistake, the server will ignore it too. 

Maybe in the future it can be proposed to add a control sequence to tell terminals the name of the currently running application, so that terminals can do more things, for example, `WindTerm` can help you to delete multi-line selections in `vi`.

**Reply by eugenov** [reply]
Thank you for clarification.
